### PR TITLE
[Docs] Add `Inferencer` development document.

### PR DESCRIPTION
**Pull request by HAOCHENYE**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: docs/zh_cn/design/infer.md [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+
+返回值：
+
+- 可视化结果和预测结果，类型为一个字典。OpenMMLab 系列算法库要求返回的字典包含 `predictions` 和 `visualization` 两个 key。
+
+### `__call__()`
+
+入参：
+
+- inputs：输入数据，通常为图片路径、或者图片数据组成的列表
```
Comment (ZwwWayne) [inline docs/zh_cn/design/infer.md:232]: 如果输入为多个内容的组合，例如：文字+图片，点云+图片，该如何处理，需要作说明

File: docs/zh_cn/design/infer.md [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+@INFERENCERS.register_module()
+class CustomInferencer(BaseInferencer):
+    ...
+```
+
+**核心接口说明**：
+
+### `__init__()`
+
+`BaseInferencer.__init__` 已经实现了[使用样例](#构建推理器)中构建推理器的逻辑，因此通常情况下不需要重写 `__init__` 函数。如果想实现自定义的加载配置文件、权重初始化、pipeline 初始化等逻辑，也可以重写 `__init__` 函数。
```
Comment (zhouzaida) [inline docs/zh_cn/design/infer.md:149]:
```suggestion
`BaseInferencer.__init__` 已经实现了[使用样例](#构建推理器)中构建推理器的逻辑，因此通常情况下不需要重写 `__init__` 函数。如果想实现自定义的加载配置文件、权重初始化、pipeline 初始化等逻辑，也可以重写 `__init__` 方法。
```

File: docs/zh_cn/design/infer.md [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+# 输入为图片路径
+img = 'path/to/img.jpg'
+result = inferencer(img)
+
+# 输入为读取的图片(类型为 np.ndarray)
+img = cv2.imread('path/to/img.jpg')
+result = inferencer(img)
+
+# 输入为 url
+
```
Comment (zhouzaida) [inline docs/zh_cn/design/infer.md:71]:
```suggestion
```

File: docs/zh_cn/design/infer.md [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+inferencer 执行推理时，通常会执行以下步骤：
+
+1. preprocess：输入数据预处理，包括数据读取、数据预处理、数据格式转换等
+2. forward: 模型前向推理
+3. visualize：预测结果可视化
+4. postprocess：预测结果后处理，包括结果格式转换、导出预测结果等
+
+为了优化 inferencer 的使用体验，我们不希望使用者在执行推理时，需要为每个过程都配置一遍参数。换句话说，我们希望使用者可以在不感知上述流程的情况下，简单为 `__call__` 接口配置参数，即可完成推理。
+
+`__call__` 接口会按照顺序执行上述步骤，但是本身却不知道使用者传入的参数需要分发给哪个步骤，因此开发者在实现 `CustomInferencer` 时，需要定义 `preprocess_kwargs`，`forward_kwargs`，`visualize_kwargs`，`postprocess_kwargs` 4 个类属性，每个属性均为一个字符元组（`Set[str]`），用于指定 `__call__` 接口中的参数对应哪个步骤：
```
Comment (zhouzaida) [inline docs/zh_cn/design/infer.md:98]:
```suggestion
`__call__` 接口会按照顺序执行上述步骤，但是本身却不知道使用者传入的参数需要分发给哪个步骤，因此开发者在实现 `CustomInferencer` 时，需要定义 `preprocess_kwargs`，`forward_kwargs`，`visualize_kwargs`，`postprocess_kwargs` 4 个类属性，每个属性均为一个字符集合（`Set[str]`），用于指定 `__call__` 接口中的参数对应哪个步骤：
```

File: docs/zh_cn/design/infer.md [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+
+`BaseInferencer.__init__` 已经实现了[使用样例](#构建推理器)中构建推理器的逻辑，因此通常情况下不需要重写 `__init__` 函数。如果想实现自定义的加载配置文件、权重初始化、pipeline 初始化等逻辑，也可以重写 `__init__` 函数。
+
+### `_init_pipeline()`
+
+```{note}
+抽象方法，子类必须实现
+```
+
+初始化并返回 inferencer 所需的 pipeline。pipeline 用于单张图片，类似于 OpenMMLab 系列算法库中定义的 `train_pipeline`，`test_pipeline`。使用者调用 `__call__` 接口传入的每个 `inputs`，都会经过 pipeline 处理，组成 batch data，然后传入 `forward` 函数。
```
Comment (zhouzaida) [inline docs/zh_cn/design/infer.md:157]:
```suggestion
初始化并返回 inferencer 所需的 pipeline。pipeline 用于单张图片，类似于 OpenMMLab 系列算法库中定义的 `train_pipeline`，`test_pipeline`。使用者调用 `__call__` 接口传入的每个 `inputs`，都会经过 pipeline 处理，组成 batch data，然后传入 `forward` 方法。
```

File: docs/zh_cn/design/infer.md [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+入参：
+
+- inputs：输入数据，未经过预处理的原始数据。
+- preds：模型的预测结果
+- show：是否可视化
+- 其他参数：由用户传入，且在 `visualize_kwargs` 中指定
+
+返回值：
+
+- 可视化结果，类型通常为 `List[np.ndarray]`
```
Comment (C1rN09) [inline docs/zh_cn/design/infer.md:214]: 感觉描述比较模糊？这个可视化结果需要满足哪些要求吗，比如必须可以 imshow 之类的？

File: docs/zh_cn/design/infer.md [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+
+```{note}
+抽象方法，子类必须实现
+```
+
+入参：
+
+- preds：模型预测结果，类型为 `list`，列表中的每个元素表示一个数据的预测结果。OpenMMLab 系列算法库中，预测结果中每个元素的类型均为 `BaseDataElement`
+- visualization：可视化结果
+- return_datasample：是否将 datasample 转换成 `dict` 返回
```
Comment (C1rN09) [inline docs/zh_cn/design/infer.md:226]:
```suggestion
- return_datasample：是否维持 datasample 返回。`False` 时转换成 `dict` 返回
```

**Review by zhouzaida — APPROVED** [review]

**Pull request merged**